server/src/main/java/org/elasticsearch/repositories/IndexId.java

```diff
@@ -42,16 +42,14 @@
     private final int hashCode;
 
     public IndexId(final String name, final String id) {
-        this.name = name;
-        this.id = id;
+        this.name = name.intern();
```

Comment:
Moving to interning here and for the snapshot id actually makes the on heap `RepositoryData` about the same size as the serialized one (unfortunately for the time being we serialize it as json uncompressed) => I think the safety check for 500kb I added is valid.

Reply:
I would prefer to avoid JVM String interning. If we can deduplicate the Strings ourselves, that's fine, but adding more and more stuff to be internalized feels dangerous (another source of OOM).